src/shared/cgroup-util.c

```diff
@@ -1795,3 +1795,44 @@ int cg_kernel_controllers(Set *controllers) {
 
         return 0;
 }
+
+
+int cg_cpu_shares_parse(const char *s, uint64_t *ret) {
+        uint64_t u;
+        int r;
+
+        if (isempty(s)) {
+                *ret = CGROUP_CPU_SHARES_INVALID;
+                return 0;
+        }
+
+        r = safe_atou64(s, &u);
+        if (r < 0)
+                return r;
+
+        if (u < CGROUP_CPU_SHARES_MIN || u > CGROUP_CPU_SHARES_MAX)
```

Comment:
Checking just interval is incorrect. Bellow reproducer exhibits the issue
```sh
$ systemd-run sleep 3600
Running as unit run-50.service.
$ systemctl set-property run-50.service CPUShares=2048
$ systemctl set-property run-50.service CPUShares=-1
Failed to parse CPUShares value -1.
```

I think we should check the value using 
```C
CGROUP_CPU_SHARES_IS_OK()
```

Same applies to blkio. Same bug is also in the latest upstream version.

Reply:
The problem is not in not comparing to CGROUP_CPU_SHARES_INVALID. It actually never gets to it because of 1809:  safe_atou64(s, &u); which returns -ERANGE if string with value starts with '-'.